src/contract/variables/safeunorderedmap.h

```diff
@@ -226,69 +346,107 @@ template <typename Key, typename T> class SafeUnorderedMap : public SafeBase {
       return size_;
     }
 
-    // TODO: somehow figure out a way to make loops work with this class (for (const auto& [key, value] : map) { ... })
+    /**
+     * Check if the map is empty (has no values).
+     * @return `true` if map is empty, `false` otherwise.
+     */
+    inline bool empty() const noexcept { return size() == 0; }
+
+    ///@{
+    /**
+     * Find a given key.
+     * @param key The key to find.
+     * @return An iterator to the found key and its value.
+     */
+    typename SafeUnorderedMap<Key, T>::iterator find(const Key& key) {
+      check();
+      auto mapPtrIt = mapPtr_->find(key);
+      if (mapPtrIt != mapPtr_->end()) {
+        return iterator(this, mapPtrIt, true);
+      }
+      auto mapIt = map_.find(key);
+      if (mapIt != map_.end()) {
+        if (erasedKeys_->find(key) == erasedKeys_->end()) {
+          return iterator(this, mapIt, false);
+        }
+      }
+      return end();
+    }
+    typename SafeUnorderedMap<Key, T>::const_iterator find(const Key& key) const {
+      check();
+      auto mapPtrIt = mapPtr_->find(key);
+      if (mapPtrIt != mapPtr_->end()) {
+        return const_iterator(this, mapPtrIt, true);
+      }
+      auto mapIt = map_.find(key);
+      if (mapIt != map_.end()) {
+        if (erasedKeys_->find(key) == erasedKeys_->end()) {
+          return const_iterator(this, mapIt, false);
+        }
+      }
+      return end();
+    }
+    ///@}
 
     /**
      * Insert a value into the map.
      * @param value The value to insert.
      * @return A pair consisting of an iterator to the inserted value and a
      *         boolean indicating whether the insertion was successful.
      */
-    std::pair<typename std::unordered_map<Key, T, SafeHash>::iterator, bool> insert(
-      const typename std::unordered_map<Key, T, SafeHash>::value_type& value
+    std::pair<typename SafeUnorderedMap<Key, T>::iterator, bool> insert(
+      const typename SafeUnorderedMap<Key, T>::value_type& value
     ) {
-      check(); markAsUsed(); dirtySize_ = true; return mapPtr_->insert(value);
+      check(); markAsUsed();
+      auto r = mapPtr_->insert(value);
+      if (!dirtySize_ && r.second && !hasCommittedKey(value.first)) ++size_;
+      return {iterator(this, r.first, true), r.second};
     }
 
-    ///@{
     /**
      * Insert a value into the map, using move.
      * @param value The value to insert.
      * @return A pair consisting of an iterator to the inserted value and a
      *         boolean indicating whether the insertion was successful.
      */
-    std::pair<typename std::unordered_map<Key, T, SafeHash>::iterator, bool> insert(
-      typename std::unordered_map<Key, T, SafeHash>::value_type&& value
+    std::pair<typename SafeUnorderedMap<Key, T>::iterator, bool> insert(
+      const typename SafeUnorderedMap<Key, T>::value_type&& value
```

Comment:
`const` here is preventing moving. Remember that moving A to B requires modifying A, so it can't be `const`. The current code will actually cause a copy.